File: Sep/Day6/minimum-operations-to-make-array-elements-zero.java
//  Time: O(N * Log(R))  :: Space: O(1)
class Solution {
    private long minimumOperations(long num) {
        long binaryLength = 1;
        long base = 1;
        long count = 0;
        while (base <= num) {
            count += ((binaryLength + 1) / 2) * (Math.min(base * 2 - 1, num) - base + 1);
            binaryLength++;
            base *= 2;
        }
        return count;
    }

    public long minOperations(int[][] queries) {
        long sumOfResults = 0;
        for (int[] query : queries) {
            sumOfResults += (minimumOperations(query[1]) - minimumOperations(query[0] - 1) + 1) / 2;
        }
        return sumOfResults;
    }
}

File: Sep/Day6/minimum-operations-to-make-array-elements-zero.js
//  Time: O(N * Log(R))  :: Space: O(1)
class Solution {
    minimumOperations(num) {
        let binaryLength = 1;
        let base = 1;
        let count = 0n;  // use BigInt for safety

        num = BigInt(num);
        base = BigInt(base);

        while (base <= num) {
            count += BigInt((binaryLength + 1) >> 1) * (BigInt(Math.min(Number(base * 2n - 1n), Number(num))) - base + 1n);
            binaryLength++;
            base *= 2n;
        }
        return count;
    }

    minOperations(queries) {
        let sumOfResults = 0n;
        for (let query of queries) {
            let l = BigInt(query[0]);
            let r = BigInt(query[1]);
            sumOfResults += (this.minimumOperations(r) - this.minimumOperations(l - 1n) + 1n) / 2n;
        }
        return sumOfResults;
    }
}
